nixos/logiops: init

ckiee opened the pull request:
###### Description of changes

This adds a new NixOS module for [logiops](https://github.com/PixlOne/logiops) which you can read more about in #165936. 

It depends on #208747 / #246115 (the `libconfig` serializer) being merged first and is in draft until that.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1036

SuperSandro2000 commented on nixos/modules/hardware/logiops/default.nix:
```diff
@@ -0,0 +1,38 @@
+
+with lib;
+
+let cfg = config.services.logiops;
```
```suggestion
let
  cfg = config.services.logiops;
```

SuperSandro2000 commented on nixos/modules/hardware/logiops/default.nix:
```diff
@@ -0,0 +1,38 @@
+
+let cfg = config.services.logiops;
+in {
+  meta.maintainers = with maintainers; [ ckie ];
+
```
```suggestion
```

SuperSandro2000 commented on nixos/modules/hardware/logiops/default.nix:
```diff
@@ -0,0 +1,38 @@
+
+    systemd.packages = [ cfg.package ];
+    systemd.services.logid.wantedBy = [ "multi-user.target" ];
+  };
```
```suggestion
  };

  meta.maintainers = with maintainers; [ ckie ];
```

SuperSandro2000 commented on nixos/modules/hardware/logiops/default.nix:
```diff
@@ -0,0 +1,38 @@
+
+  config = mkIf cfg.enable {
+    services.udev.packages = [ pkgs.logitech-udev-rules ];
+    environment.etc."logid.cfg".source =
+      (pkgs.formats.libconfig { }).generate "logid.cfg" cfg.settings;
+
```
```suggestion
    environment.etc."logid.cfg".source = (pkgs.formats.libconfig { }).generate "logid.cfg" cfg.settings;

```

ckiee commented:
For some reason my Emacs is executing nixfmt instead of nixpkgs-fmt. Weird.

SuperSandro2000 reviewed (commented):
module LGTM, someone still needs to review the lib changes.

ckiee commented:
cc @Ma27 @lheckemann if either of you feel like reviewing

anund commented on nixos/modules/hardware/logiops/default.nix:
```diff
@@ -0,0 +1,39 @@
+    environment.etc."logid.cfg".source = (pkgs.formats.libconfig { }).generate "logid.cfg" cfg.settings;
+
+    systemd.packages = [ cfg.package ];
+    systemd.services.logid.wantedBy = [ "multi-user.target" ];
```
Automatically reload the service on config change:
```nix
systemd.services.logid.restartTriggers = [ config.environment.etc."logid.cfg".source ];
```

ckiee committed:
logiops: remove reference to broken reload mechanism

ckiee commented on nixos/modules/hardware/logiops/default.nix:
```diff
@@ -0,0 +1,38 @@
+    environment.etc."logid.cfg".source =
+      (pkgs.formats.libconfig { }).generate "logid.cfg" cfg.settings;
+
+    systemd.packages = [ cfg.package ];
```
The latest release has `logid::reload` commented out. The HUP isn't caught and `$MAINPID` gets killed. I'll make it complain about that instead.

SuperSandro2000 commented on nixos/modules/hardware/logiops/default.nix:
```diff
@@ -0,0 +1,79 @@
+
+let
+  cfg = config.services.logiops;
+  renderedConfig =
+    (pkgs.formats.libconfig { }).generate "logid.cfg" cfg.settings;
```
```suggestion
  renderedConfig = (pkgs.formats.libconfig { }).generate "logid.cfg" cfg.settings;
```

ckiee replied:
nixpkgs-fmt doesn't change this and it *is* a long line. Why?

SuperSandro2000 replied:
because nixpkgs-fmt is not black and won't change long lines back.

ckiee replied:
It's also over 80 columns and won't fit on my whole screen if I have another window open next to it which is a bummer.

(Anyhow, "fixed")

SuperSandro2000 commented on pkgs/misc/drivers/logiops/default.nix:
```diff
@@ -13,9 +13,14 @@ stdenv.mkDerivation rec {
   PKG_CONFIG_SYSTEMD_SYSTEMDSYSTEMUNITDIR = "${placeholder "out"}/lib/systemd/system";
 
-  nativeBuildInputs = [ cmake pkg-config ];
+  nativeBuildInputs = [ cmake pkg-config gnused ];
```
```suggestion
  nativeBuildInputs = [ cmake pkg-config ];
```
part of stdenv

SuperSandro2000 commented on pkgs/misc/drivers/logiops/default.nix:
```diff
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchFromGitHub, pkg-config, cmake, udev, libevdev, libconfig }:
+{ lib, stdenv, fetchFromGitHub, pkg-config, cmake, udev, libevdev, libconfig, gnused }:
```
```suggestion
{ lib, stdenv, fetchFromGitHub, pkg-config, cmake, udev, libevdev, libconfig }:
```

anund commented on nixos/modules/hardware/logiops/default.nix:
```diff
@@ -0,0 +1,79 @@
+
+          buttons = [
+            {
+              cid = "0x53";
```
My mistake. The literal `0x53` is necessary not the string `"0x53"`. At the moment this generates the error `[WARN] Line 16: cid must be a number, ignoring.`

anund replied:
https://hyperrealm.github.io/libconfig/libconfig_manual.html#Integer-Values Hex is part of the spec.

ckiee replied:
Applying your patch from the other thread resolves this. (also added a bit more)

```nix
nix-repl> :l .
nix-repl> :b (formats.libconfig {}).generate "test" example
https://i.ckie.dev/hPpbjIA.txt (browser default is tab width 8)

anund commented on lib/generators.nix:
```diff
@@ -334,6 +334,78 @@ rec {
 ${expr "" v}
 </plist>'';
 
+  toLibconfig = { }:
```
This can't currently generate valid config for the example. :(

anund replied:
```git
diff --git a/lib/generators.nix b/lib/generators.nix
index 0eab0c076c1..23f57e5fdcd 100644
--- a/lib/generators.nix
+++ b/lib/generators.nix
@@ -393,6 +393,39 @@ ${expr "" v}
     v:
     let
       isFloat = builtins.isFloat or (x: false);
+      isHexString = str: builtins.match "^[+-]?0x[0-9a-fA-F]+$" str != null;
+      isOctalString = str: builtins.match "^[+-]?0[1-7][0-7]*$" str != null;
+      isLongString = str: builtins.match "^[+-]?[0-9]+L$" str != null;
+      isNumber = x: (isHexString x) || (isOctalString x) || (isLongString x);
+      isScalar = builtins.isString or isNumber or builtins.isFloat or builtins.isBool;
+
+      isArray = x: let
+        typed = builtins.groupBy type x;
+      in
+        !(typed ? "null") && !(typed ? "complex");
+
+      type = x:
+        with builtins;
+        if x == null then
+          "null"
+        else if isBool x then
+          "bool"
+        else if isInt x then
+          "int"
+        else if isFloat x then
+          "float"
+        else if isString x then
+          if isHexString x then
+            "hex"
+          else if isOctalString x then
+            "octal"
+          else if isLongString x then
+            "long"
+          else
+            "string"
+        else
+          "complex";
+
       expr = ind: x:
         with builtins;
         if x == null then
@@ -404,7 +437,10 @@ ${expr "" v}
         else if isFloat x then
           num ind x
         else if isString x then
-          str ind x
+          if isNumber x then
+            lit ind x
+          else
+            str ind x
         else if isPath x then
           path ind x
         else if isList x then
@@ -421,6 +457,7 @@ ${expr "" v}
       str = ind: x: literal ind ''"${x}"'';
       key = ind: x: literal ind "${x}: ";
       path = ind: x: literal ind (toString x);
+      lit = ind: x: literal ind "${x}";
 
       indent = ind: expr "\t${ind}";
 
@@ -430,11 +467,18 @@ ${expr "" v}
         '' (indent ind);
 
       list = ind: x:
-        libStr.concatStringsSep "\n" [
-          (literal ind "(")
-          (item ind x)
-          (literal ind ")")
-        ];
+        if isArray x then
+          libStr.concatStringsSep "\n" [
+            (literal ind "[")
+            (item ind x)
+            (literal ind "]")
+          ]
+        else
+          libStr.concatStringsSep "\n" [
+            (literal ind "(")
+            (item ind x)
+            (literal ind ")")
+          ];
 
       attrs = ind: x:
         libStr.concatStringsSep "\n" [
```

Here's a vaguely correct patch to allow the example to generate valid config. [Integers](https://hyperrealm.github.io/libconfig/libconfig_manual.html#Integer-Values), and [Longs](https://hyperrealm.github.io/libconfig/libconfig_manual.html#g_t64_002dbit-Integer-Values) need you to squint at strings to see if they are really number literals. (regex... uh... works). [Arrays](https://hyperrealm.github.io/libconfig/libconfig_manual.html#Arrays) are just ambiguous in nix config. The above "works". I'd be worried `(...)` where all elements had the same type is expected somewhere and the above would spit out `[...]` instead breaking some tool. Explaining some conventional naming strategy  (`key.array = [...]`) would work but be confusing to debug.

This seems like a place nix the language isn't suitable on it's own. Alternatives?

ckiee replied:
I can't do stuff right now but I've been using this with [my Nix module](https://github.com/ckiee/nixfiles/blob/0ac57655f15500051736923287ca55f3c84b65e5/modules/logiops.nix#L14) and it's been working quite nicely for a while (excluding issues on wake from suspend occasionally). Here's the [`logid.conf`](https://i.ckie.dev/J0qoyzt.cfg).

> This can't currently generate valid config for the example. :(

It should be able to if you dare to manually convert hex to decimal (as seen in my link above)

> The libconfig implementation won't be good enough for users.

It's been good enough for me and that's why I'm here to share it. I'll apply your patch soon but I wrote just enough to have it work for me and I do not care about spec compliance as long as I'm not doing [anything surprising](https://en.wikipedia.org/wiki/Principle_of_least_astonishment). 

> This seems like a place nix the language isn't suitable on it's own. Alternatives?

Then I can't just do this:

```
ckie@cookiemonster ~/git/nixfiles -> c eval 'nodes.'$(hostname)'.config.services.logiops.settings.devices' | head -n5
[
  {
    "buttons": [
      {
        "action": {
```

Or override a value for a specific system.

anund replied:
[@include](https://hyperrealm.github.io/libconfig/libconfig_manual.html#Include-Directives) should also be supportable with `"@include" = "...".

anund replied:
I don't mean to imply what you've done isn't helpful. To be clear the example that I previously suggested will not work (`"cid": "0x56"`).

Regarding nix: this comment is too brief on my part. I'm doing parsing work with regex trying to apply types to strings. It works but likely could be better with different tools :shrug:. I also should not have implied I wanted you in particular to come up with an alternative.

Regarding surprise: I tried the least surprising thing to me :D. I produced an example using strings for hexadecimal values. Likely assuming `"..."` would be passed through without quotes as I wasn't aware of the types used by libconfig. Similarly I'm surprised `()` and `[]` are interchangeable in `keys:`. Poking at the spec I can guess why it would work.

Most of what I've attempted to address, and poorly communicated, is treating `lib.toLibconfig` as a workable tool for others. Your work around for the integer type is probably all that's needed for `logiops` and can be addressed in the `example`, `description`, and perhaps a comment on the generator.

ckiee replied:
> I also should not have implied I wanted you in particular to come up with an alternative.

I think not sleeping for >15.5 hours probably had something to do with my rather harsh reply. Sorry :P

Anyway I applied your patch and looked at the spec and it's actually pretty small which makes me understand why you wanted it to be complaint. I guess I just don't like the "user" phrasing. Sorry again.

anund reviewed (changes requested):
The libconfig implementation won't be good enough for users.

ckiee commented:
Oops. Put the new libconfig changes in the wrong commit.

ckiee committed:
nixos/logiops: init

anund commented:
If the creating a lib package is stalling this PR should the changes to lib be moved into the module? It's not clear if there's anyone to actually review the lib part of the change.

ckiee committed:
nixosTests.logiops: add

There's no negative test due to runner limitations.

I tried to hack on the runner for a bit but it got out of scope and I'm
not interested in learning more Python. My 'fork' of nixpkgs has a
logiops-nixos-xtra branch for anyone interested in poking the testing
more.

ckiee commented:
I added a NixOS test to hopefully quench worries, but it doesn't actually pass. Logiops is accepting a completely invalid configuration: ([log of build of `pn0zlc8089325nfnvfjs3z2ni0lml4gw-vm-test-run-logiops.drv`](https://i.ckie.dev/1VqQyqT.txt
), better to `curl` it for ANSI colors)

cc @DavHau if you wanna review the lib part ([reason](https://github.com/DavHau/fromYaml))

anund commented on lib/generators.nix:
```diff
@@ -334,6 +334,130 @@ rec {
+      isOctalString = str: builtins.match "^[+-]?0[1-7][0-7]*$" str != null;
+      isLongString = str: builtins.match "^[+-]?[0-9]+L$" str != null;
+      isNumber = x: (isHexString x) || (isOctalString x) || (isLongString x);
+      isScalar = builtins.isString or isNumber or builtins.isFloat or builtins.isBool;
```
oops, this line isn't used

ckiee committed:
[PR dependency end]

h7x4 committed:
formats.libconfig: init

Co-authored-by: ckie <25263210+ckiee@users.noreply.github.com>
Signed-off-by: h7x4 <h7x4@nani.wtf>

h7x4 committed:
formats.libconfig: add tests

Co-authored-by: ckie <25263210+ckiee@users.noreply.github.com>
Signed-off-by: h7x4 <h7x4@nani.wtf>

ckiee committed:
pkgs-lib.formats: add note about missing `pkgs` quirk

ckiee commented:
rebased on everything and attempted to bump logiops but it is very broken (PixlOne/logiops#409, among other possible issues) so that work [is in a different branch](https://github.com/ckiee/nixpkgs/tree/logiops-bump-023-033).
